alf/algorithms/sac_algorithm.py

```diff
@@ -290,32 +287,71 @@ def actor_loss_fn(dqda, action):
             log_action_probs = base_action_dist.logits.squeeze(1)
 
             target_q_value = critics.min(dim=1)[0].detach()
-            alpha = torch.exp(self._log_alpha)
             actor_loss = torch.exp(log_action_probs) * (
-                alpha.detach() * log_action_probs - target_q_value)
+                alpha * log_action_probs - target_q_value)
             actor_loss = actor_loss.mean(list(range(1, actor_loss.ndim)))
+        else:
+            # With CriticNetwork, we can only make one action's probability
+            # proportional to the corresponding Q value at a time.
+            # If continuous, use the original SAC's formulation; otherwise only
+            # train one entry of a categorical distribution.
+            # In either case, the target Q value to be fit to is the same.
+            target_q_value = critics.min(dim=1)[0]
+            cont_act_logpi = nest.flatten_up_to(
+                self._is_continuous,
+                nest.map_structure(lambda *args: tuple(args),
+                                   self._is_continuous, action, log_pi))
+
+            def _train_continuous_actor(action, log_pi):
+                dqda = list(torch.autograd.grad(target_q_value.sum(), action))
+
+                def actor_loss_fn(dqda, action, log_pi):
+                    if self._dqda_clipping:
+                        dqda = torch.clamp(dqda, -self._dqda_clipping,
+                                           self._dqda_clipping)
+                    loss = 0.5 * losses.element_wise_squared_loss(
+                        (dqda + action).detach(), action)
+                    loss = loss.sum(list(range(1, loss.ndim)))
+                    loss += alpha * log_pi
+                    return loss
+
+                return nest.map_structure(actor_loss_fn, dqda, action, log_pi)
+
+            def _train_discrete_actor(log_pi):
+                def actor_loss_fn(log_pi):
+                    # See ``notes/estimating_derivative_of_expectation.rst``
+                    # for how this loss formula is derived.
+                    loss = log_pi * (alpha * log_pi - target_q_value).detach()
```

Comment:
Similar to policy gradient, subtracting a base might be helpful for convergence because of lower variance